File: wasm-wrapper/bech32.go
//go:build js && wasm

package main

import (
	"encoding/hex"
	"syscall/js"

	bech32pkg "github.com/btcsuite/btcd/btcutil/bech32"
)

func bech32Encode(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 2, "hrp, hexData5bit"); e != nil {
		return e
	}
	data, e := hexDecode(args[1].String())
	if e != nil {
		return e
	}
	s, err := bech32pkg.Encode(args[0].String(), data)
	if err != nil {
		return errfResult("bech32 encode: %s", err)
	}
	return okResult(s)
}

func bech32EncodeM(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 2, "hrp, hexData5bit"); e != nil {
		return e
	}
	data, e := hexDecode(args[1].String())
	if e != nil {
		return e
	}
	s, err := bech32pkg.EncodeM(args[0].String(), data)
	if err != nil {
		return errfResult("bech32m encode: %s", err)
	}
	return okResult(s)
}

func bech32Decode(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 1, "bechStr"); e != nil {
		return e
	}
	hrp, data, err := bech32pkg.Decode(args[0].String())
	if err != nil {
		return errfResult("bech32 decode: %s", err)
	}
	return okResult(map[string]any{
		"hrp":  hrp,
		"data": hex.EncodeToString(data),
	})
}

func bech32DecodeNoLimit(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 1, "bechStr"); e != nil {
		return e
	}
	hrp, data, err := bech32pkg.DecodeNoLimit(args[0].String())
	if err != nil {
		return errfResult("bech32 decodeNoLimit: %s", err)
	}
	return okResult(map[string]any{
		"hrp":  hrp,
		"data": hex.EncodeToString(data),
	})
}

func bech32EncodeFromBase256(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 2, "hrp, hexData"); e != nil {
		return e
	}
	data, e := hexDecode(args[1].String())
	if e != nil {
		return e
	}
	s, err := bech32pkg.EncodeFromBase256(args[0].String(), data)
	if err != nil {
		return errfResult("bech32 encodeFromBase256: %s", err)
	}
	return okResult(s)
}

func bech32DecodeToBase256(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 1, "bechStr"); e != nil {
		return e
	}
	hrp, data, err := bech32pkg.DecodeToBase256(args[0].String())
	if err != nil {
		return errfResult("bech32 decodeToBase256: %s", err)
	}
	return okResult(map[string]any{
		"hrp":  hrp,
		"data": hex.EncodeToString(data),
	})
}

func bech32ConvertBits(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 4, "hexData, fromBits, toBits, pad"); e != nil {
		return e
	}
	data, e := hexDecode(args[0].String())
	if e != nil {
		return e
	}
	result, err := bech32pkg.ConvertBits(
		data,
		uint8(args[1].Int()),
		uint8(args[2].Int()),
		args[3].Bool(),
	)
	if err != nil {
		return errfResult("convertBits: %s", err)
	}
	return okResult(hex.EncodeToString(result))
}

File: src/bech32.ts
import { init, g, unwrap } from './init';
import type { Bech32DecodeResult, Bytes } from './types';

/** Bech32 and bech32m encoding/decoding. */
export const bech32 = {
  /** Bech32-encode 5-bit data (hex) with the given HRP.
   *  Calls Go: bech32.Encode() from btcutil/bech32. */
  async encode(hrp: string, data5bit: Bytes): Promise<string> {
    await init();
    return unwrap<string>(g().bech32.encode(hrp, data5bit));
  },

  /** Bech32m-encode 5-bit data (hex) with the given HRP.
   *  Calls Go: bech32.EncodeM() from btcutil/bech32. */
  async encodeM(hrp: string, data5bit: Bytes): Promise<string> {
    await init();
    return unwrap<string>(g().bech32.encodeM(hrp, data5bit));
  },

  /** Decode a bech32 string (90-char limit). Returns 5-bit data.
   *  Calls Go: bech32.Decode() from btcutil/bech32. */
  async decode(str: string): Promise<Bech32DecodeResult> {
    await init();
    return unwrap<Bech32DecodeResult>(g().bech32.decode(str));
  },

  /** Decode a bech32 string without length limit. Returns 5-bit data.
   *  Calls Go: bech32.DecodeNoLimit() from btcutil/bech32. */
  async decodeNoLimit(str: string): Promise<Bech32DecodeResult> {
    await init();
    return unwrap<Bech32DecodeResult>(g().bech32.decodeNoLimit(str));
  },

  /** Encode base-256 data (hex) as a bech32 string. Handles 8->5 bit conversion.
   *  Calls Go: bech32.EncodeFromBase256() from btcutil/bech32. */
  async encodeFromBase256(hrp: string, data: Bytes): Promise<string> {
    await init();
    return unwrap<string>(g().bech32.encodeFromBase256(hrp, data));
  },

  /** Decode a bech32 string to base-256 data (hex). Handles 5->8 bit conversion.
   *  Calls Go: bech32.DecodeToBase256() from btcutil/bech32. */
  async decodeToBase256(str: string): Promise<Bech32DecodeResult> {
    await init();
    return unwrap<Bech32DecodeResult>(g().bech32.decodeToBase256(str));
  },

  /** Convert between bit groups (e.g. 8->5 or 5->8).
   *  Calls Go: bech32.ConvertBits() from btcutil/bech32. */
  async convertBits(
    data: Bytes,
    fromBits: number,
    toBits: number,
    pad: boolean,
  ): Promise<Uint8Array> {
    await init();
    return unwrap<Uint8Array>(
      g().bech32.convertBits(data, fromBits, toBits, pad),
    );
  },
};
